### DISCUSSION TITLE
Using the Explainer model with our own  models

### DISCUSSION THREAD

**Discussion by alzaia** [pyg-team/pytorch_geometric#5854] [Q&A] [answered]
Hello, I'm having a bit of trouble using the `Explainer` (e.g. `GNNExplainer` model) on top of my PyG homemade GNN model. The issue is that the `Explainer` assumes a certain input args in the forward method of the model:
https://github.com/pyg-team/pytorch_geometric/blob/48ae5057653ec69debbe03927e777005fc268b4e/torch_geometric/nn/models/explainer.py#L237
But the args in my user-defined model are different (I just pass the `batch` and then use whatever I need from that (e.g. `batch.x`, `batch.edge_index`, `batch.edge_attr`, ...). I assume this can also cause issues if used with some of your pre-defined models. For example:
https://github.com/pyg-team/pytorch_geometric/blob/48ae5057653ec69debbe03927e777005fc268b4e/torch_geometric/nn/models/lightgcn.py#L101

Is this a limitation with the current implementation of the `Explainer`?

**Comment by alzaia** [comment, marked as answer]
I guess this is directly related to this discussion: https://github.com/pyg-team/pytorch_geometric/discussions/5814

**Reply by rusty1s** [reply]
Yes, this is a current limitation. We can either use the first two arguments to expect node features and edge indices. Alternatively, we can fix name convention such that `x` and `edge_index` needs to be present in the input arguments but in any order.

I think it is generally bad practice to use `Data` as input to `model.forward` though, especially since it is not well-defined which attributes in `Data` exist to make your model run successfully. I think something like
```python
def forward(self, x, edge_index, ...):
    pass

model(data.x, data.edge_index)
```
is easier to understand.

**Reply by alzaia** [reply]
Thanks a lot for the feedback!
